### PR TITLE
pattern/open-source-trumps-innersource

### PR DESCRIPTION
**Pull request by InnerSrcAdmin**
People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.

NEXT STEP: Needs to be revised and reviewed
NOTES: Needs further clarification - 2016-10-27

### CODE REVIEW

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: What does _BL_ stand for? Is it business line? Is this different from a _BU_ (business unit)?

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: I personally don't think this is a strong force. IMHO, voluntariness pertains more to the development of and making contributions to OSS, not necessarily to using it.

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: One more possible force in favour of using InnerSource components: licensing. Depending on the regulations within a company, it might be easier to use a _corporate licensed_, InnerSource component than an open source component which is licensed in a way that requires the company to disclose IP.

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: Might this be a problem for a dedicated pattern/donut? I feel this is not so much related to OSS trumping InnerSource.

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: I personally feel that assigning maintainers is the key solution here. I think it deserves a bullet point of its own.

**Review by gruetter — CHANGES_REQUESTED** [review]

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: BL = Business Line. Typically there are multiple BLs in a Business Unit; but I can guess this varies depending on company.

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: Interested in a **Patlet** in this pattern?

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: It's not clear to me that this behavior is a problem.

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: This comment doesn't have to be addressed right now, but in the log run it would be nice if our patterns had consistent terminology (glossary) for terms like these (`Business Lines`).

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: This line seems like part-`Context` (`Internally developed components should have advantages over competing open source`) and part-`Solution` (`If you can clearly demonstrate this, it can be persuasive.`) rather than `Forces`.

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: If this is a problem then put it in the `Problem` section.

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: Use more clear, definite language (rather than conditional).

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: It's unclear to me what these two last bullet points are trying to get at or how they are uniquely appropriate to describing this pattern?

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: If I understand the diagram (and this article) right, then it looks like the diagram suggests that it is preferable to use an inner source project over an open source project in areas of software that are differentiating to the company's core business?

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: I doubt the ability for such a requirement to have real, lasting, and actual affect on employee behavior.

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: Quite a bit of overlap with `Create a tool with a central view of internally available software`.  Maybe just delete this sentence?

File: open-source-trumps-innersource.md
Comment (rrrutledge) [inline open-source-trumps-innersource.md]: A lot of these suggested resolutions are (gut-feel) very large undertakings and outside the realm of reasonableness to just go and do.  I agree that they would help the situation your describing, but feel that they are large to the point of their not having practical utility in including them here as independent items.  Perhaps some of them need to be their own pattern and linked to from here?  Or perhaps the need to include such large items as a solution indicates that the problem space that this pattern is trying to address is too broad?

**Review by rrrutledge — DISMISSED** [review]
These thoughts are great, Nick.  I like the broad themes and big-picture items that you've outlined here and am excited to discuss more of the details.

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: Yes, it would be good to add:
## Patlet
People find the InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with the open source component.

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: I think the core problem is that people's mindset is that open source is always superior to InnerSource, which often is not the case. The result is that this bias leads people to select and use inferior components, perhaps having to put effort to adapting them that would otherwise go towards improving and reusing the InnerSource component.

Comment (rrrutledge) [reply open-source-trumps-innersource.md]: I see.  I think that point would be useful to roll into the statement - the inner source component meets their needs better than an open source component.  They should choose the component that best meets their needs irrespective of it being open source vs. inner source?

Comment (gruetter) [reply open-source-trumps-innersource.md]: As a tangent: we've seen projects chose expensive COTS over superior InnerSource software, too.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: An additional effect possibly worth mentioning is that this will also result in more effort for procurement and supplier management.

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: I'm not sold on the _channels-are-more-obvious_ argument. In fact, if both the open source and InnerSource project do an equally good job of explaining how to collaborate with them, the InnerSource project might have an advantage (e. g. being able to see the shared calendar of the maintainers).

Comment (gruetter) [reply open-source-trumps-innersource.md]: Also, would it make sense to elaborate on where the perception of higher quality comes from? In german, we have a saying that goes something like this: "The prophet doesn't count in is own country".

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: Proposal: _Using corporate licensed InnerSource components instead of using open source components with strong copyleft license avoids having to disclose IP_.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: I have the same problem.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: I have the same reservation.

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: I think this needs to be qualified. It the OSS component has strong copyleft, it might make sense to re-engineer its relevant parts in an IS project.

**Comment by spier** [comment]
Hi @NewMexicoKid @gruetter @rrrutledge (mentioning you as you spent quite some time on this PR already). 

We have reworked the [maturity levels][levels] of the InnerSource patterns. The new levels make it more explicit what quality standards are expected for each level. With that we will also try to get PRs merged into master faster.

**Status of this pattern:**

We have currently categorized the Pattern in this PR as maturity level **Level 2 - Structured**.

For that level we have two requirements:

* is validated by at least one known instance
   * we suspect that this is already true, as this pattern was previously listed in the group `Pattern Drafts (proven, not yet fully reviewed)`
* complies with the [patterns format][format]
   * not sure about this, especially as the pattern format may have evolved since this PR was opened

**How to get this PR merged?**

* Could you confirm if the status assessment above sounds correct to you?
* If this pattern does not match the [patterns format][format] yet, could you adapt it accordingly?
* With that we should be able to merge this PR to master. That will also allow other contributors to level up this pattern further in subsequent PRs/contributions in the future.

**I have no time to work on this, what to do?**

As this PR is open for a while, it would be completely understandable if your priorities have shifted and you don't time to work on this anymore.

In that case please just let us know in a comment below. Then we can decide what to do with this PR on our own.

Thank you for your help! 👍 

[levels]: https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/meta/contributor-handbook.md#maturity-levels
[format]: https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/meta/pattern-template.md

**Comment by lenucksi** [comment]
Thanks @spier for adding this update.
Since this pull request has seen a lot of interesting discussions and still has them open and unresolved I would be happy if the persons involved in these discussions could take a look to see if they are resolvable, e.g. by commit suggestions or resolve them if they deem them resolved such that we'd not loose insights by merging with open discussions.
I think @NewMexicoKid @gruetter and @rrrutledge might be interested here.

**Comment by rrrutledge** [comment]
Thanks for this message, @lenucksi.  I'm un-opinionated about what happens to my comments :)

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: @rrrutledge how about: Internally developed components can be made more robust and better suited towards corporate needs. Reasons for preferring inner source over open source alternatives might include issues with the open source license, use of proprietary and/or patented components, and functional/competitive advantages. The more that such internal components are used within a company, the better the ROI for the inner source development efforts. On the other hand, open source advances can sometimes leap frog internal development and without periodic assessment, it can be difficult to know when to stop investment in inner source development and switch to the adoption of open source components.

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: Yes

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: I think the point of this force is that it can be difficult to fight perception among dispersed and silo'd engineering staff and to communicate (1) that the inner source project exits, (2) that it has advantages over competing open source, (3) that there are legitimate reasons for preferring the inner source over competing open source software.

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: It may depend on the corporate culture or how this requirement is put in place (there might be a process for finding and evaluating software components or for using open source software). I think it is doable to require this step (we use it in our company).

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [inline open-source-trumps-innersource.md]: Suggest: Developers select the best component regardless of whether it is open source or internally developed, but they do so with full knowledge of the inner source component, its capabilities, and reasons why it is recommended. Internally developed software components are then more widely and consistently used within the company.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: I think this is not the problem this pattern addresses. I will add an additional force for this.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: If this is the core solution of this pattern, then I propose we add language to the context which makes it clear that this solution could be successfully applied.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: I'll add a link to the _InnerSource Portal_ pattern, instead

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: On 2nd thought, I think we should remove this bullet point, as it doesn't really solve the underlying problem.

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: > "formal repo solution"

Am I right assuming that you mean the mandate by governance to evaluate InnerSource components, first, @NewMexicoKid ? If so, I propose to change to

"Make InnerSource components more attractive by providing "value add" services, such as ...

File: open-source-trumps-innersource.md
Comment (gruetter) [inline open-source-trumps-innersource.md]: I feel this is a bit of an overlap to the 2nd bullet point here (Make it easier to find ...). I propose to merge these two bullet points into one and link to the "InnerSource Portal" pattern.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: Taken care of in line 22, already.

File: open-source-trumps-innersource.md
Comment (gruetter) [reply open-source-trumps-innersource.md]: @rrrutledge : Are you ok with merging with the current state of the resolution section (there has been a slight rework), or do you think further rework is needed before merging?

**Review by gruetter — APPROVED** [review]
Pattern looks good now and ready to merge, IMHO. I have not yet heard of a company which successfully adopted the core of the solution (mandate considering IS SW before OSS). The pattern conforms to the common structure, though. Therefore, I think the proper maturity level is _2-structured_, @spier , @lenucksi . I'll update the PR tag accordingly.

File: open-source-trumps-innersource.md
Comment (NewMexicoKid) [reply open-source-trumps-innersource.md]: The perception usually comes from a lack of information, a bias in believing that open source = many more developers and users = higher quality (this is sometimes the case but not always).

File: open-source-trumps-innersource.md
Comment (lenucksi) [reply open-source-trumps-innersource.md]: Could any of you two (@gruetter @NewMexicoKid) suggest any possible commit suggestion to modify this or resolve this discussion?

File: open-source-trumps-innersource.md
Comment (lenucksi) [reply open-source-trumps-innersource.md]: What do you think about the following:
- Add this .jpg to the repo maybe in /assets/ or something similar such that it can be rendered with the pattern? Also, who created the image and are there possibly any copyrights attached to it that we need to take into account?
- Should there be some phrasing adding the insight from the image into the text?

File: open-source-trumps-innersource.md
Comment (lenucksi) [inline open-source-trumps-innersource.md]: We've had quite a few discussions and edits around this. Are there suggestions on who should be included as co-authors to this?

File: open-source-trumps-innersource.md
Comment (lenucksi) [reply open-source-trumps-innersource.md]: I've referenced @rrrutledge point into #170.

I've seen @gruetter and @NewMexicoKid agreeing on the possible need to mention the procurement and supplier management point. Can any of you two create a commit suggestion or a commit around this?

File: open-source-trumps-innersource.md
Comment (lenucksi) [inline open-source-trumps-innersource.md]: I've slightly changed the phrasing here to make it easier to understand. This reflects my understanding of the rest of the pattern. Please correct if I got the point of the pattern wrong.

```suggestion
People find an InnerSource project but, after all things are considered, even if the InnerSource component meets their needs, they still go with an equivalent open source component.
```

File: open-source-trumps-innersource.md
Comment (lenucksi) [inline open-source-trumps-innersource.md]: This is pretty much equivalent to the **Patlet** can we deduplicate or remove parts here?

File: open-source-trumps-innersource.md
Comment (lenucksi) [inline open-source-trumps-innersource.md]: I think this could be phrased clearer, I do not get the full point here. Do you refer to list of suggested OSS / InnerSource solutions? Possibly along the lines of an internal tech radar? Should the external (publicly visible?) GitHub org also list InnerSource solutions?

**Review by lenucksi — COMMENTED** [review]
Thanks a lot for the review and work towards resolving open discussions @gruetter, @NewMexicoKid and @rrrutledge!

> Pattern looks good now and ready to merge, IMHO. I have not yet heard of a company which successfully adopted the core of the solution (mandate considering IS SW before OSS). The pattern conforms to the common structure, though. Therefore, I think the proper maturity level is 2-structured, @spier , @lenucksi . I'll update the PR tag accordingly.

The contribution handbook has this as a verification requirement for `2-structured`: 

> Is validated by at least one known instance
>  - Alternatively: key elements of the pattern have been validated in separate contexts and, in consequence, it is justified to believe the full solution will function

What are the key elements that have been validated in separate contexts here if there is no successful instance of deployment?

I've also added a few review comments and commit suggestion after reading the current result. This is a very interesting pattern. 😄 

And as a last element: Independent of this being merged as `2-structured` or `1-initial` we need a follow-up pull request changing moving the file to the according directory structure.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: That's also how I understand this pattern. Adding your suggestion to the PR.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: The BL acronym is spelled out in the Context section now. Resolving.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: This has been worked into the pattern already. Resolving.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: This feedback has been worked into the "Problem" section already. Resolving.

@gruetter if you would like to see your statement about COTS integrated into the pattern, would you mind sending a suggestion for this?

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: Keeping this comment open, to make it easier to find this comment after the PR is merged.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: I cannot tell if this thread has been worked into the latest version of the pattern already. Will leave the tread open.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: I see this sentence in Context which sounds similar:
> * Different Business Lines (BLs) are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.

Therefore I am assuming that this thread has been resolved.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: The last statement has been worked into the pattern already. Resolving thread.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: The "owner" of the picture in gDrive is @NewMexicoKid and it was created 11.10.2016, 18:33.
Not sure if that helps us to figure out where it originally came frame?

I also did a google image search for the picture and parts of the text, but could find anything.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: Context contains:
> * The company culture is based on strong governance.

So I think this thread is addressed. Resolving.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: Thread is worked into the pattern already. Resolving.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: While I agree that the **Solutions** section is asking for a lot right now, I suggest that we just leave this as is and get the pattern published as Initial. That will make it easier for others to find, read, and improve this pattern in the future if anybody has actually found a working solution for this problem. (right now we don't have a Known Instance yet, so this isn't a proven solution yet)

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: I worked in @NewMexicoKid comment as such:
> There is a perception that the open source components are of higher quality and more reusable than the InnerSource components. This is caused by the common belief that open source always means many more developers and users and therefore higher quality. This is sometimes the case in open source but not always!

Comment (spier) [reply open-source-trumps-innersource.md]: While I am not sold on the _channels-are-more-obvious_ argument either, I would just leave it as is for now and see what happens when we get feedback for the pattern from a wider audience. 

Resolving this thread.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: This sentence has already been changed to:
> Increase the attractiveness of InnerSource components by providing “value add” services for them, such as automated CI/CD services, IaaS/PaaS, NPM organization/server linkages, ChatOps, etc.

Resolving.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: Agreed. I have removed the point
> * Provide an internal instance of GitHub Enterprise or a well publicized external GitHub organization repo to allow employees to easily find internally developed solutions.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: While I don't fully understand what "leading repos" means in this context, @gruetter's feedback has already been incorporated as this sentence is its own bullet now:
> * Assign maintainers to make sure proper open source processes are being followed within the leading repos.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: You meant that this one could go to it's own pattern instead, right?
> Different Business Lines (BLs) are selecting different SW components to do the same function, resulting in inconsistencies in the user experience.

Will leave this thread open but I think it should not prevent us from merging this PR.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: I added all of you as. you seemed to have brought this pattern to where it is today.

new author section
```
* As told to Padma Sudarsan, 2016-09-30
* Tim Yao
* Georg Gruetter (Robert Bosch GmbH)
* Russ Rutledge
* Johannes Tigges
```

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: I removed the duplication by writing a new patlet that contains both problem and solution.
Resolving.

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: Based on a different comment I removed the "Provide an internal instance of GitHub Enterprise ..." bullet as it seemed redundant with the "Make it easier to find ..." one.
Resolving.

**Comment by spier** [comment]
I have reviewed this PR in detail, following this approach:

* Compared all comment threads with the current content of the pattern, and resolved the ones that had already been incorporated.
* For the remaining threads I worked in the feedback to the best of my ability and resolved those threads as well.
* If I didn't feel comfortable making a judgement call I left the treads open.

I opted to move this pattern to `1-initial` because it doesn't have a **Known Instance** as @lenucksi pointed out. Also content-wise it is more an initial idea than a fully thought out pattern.

I think this pattern has higher chances of improving by getting more readership, which is easier to achieve once merged. 

Therefore my suggestion for getting this PR merged are this:

* give this a couple of days for other commenters on this thread to review the latest version
* integrate the sketch image into the pattern according to [this comment](https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/46#discussion_r492680017) 
* leave the remaining threads open, even when merging this PR, in case anybody wants to look something up.

Follow a _done is better than perfect_ approach, I will wait **until Friday 2021-02-05**. Then I will work in any new feedback and get this merged as outlined above.

Any feedback from others is welcome, as always! :)

File: open-source-trumps-innersource.md
Comment (spier) [reply open-source-trumps-innersource.md]: I found the original source for this image. Yay :)
https://www.semanticscholar.org/paper/Commodification-of-Industrial-Software%3A-A-Case-for-Linden-Lundell/54d6cb77a86e292ff1845eb910c1a1f258e6cee3

**Comment by spier** [comment]
Timer is up :)

Found the [original image](https://www.semanticscholar.org/paper/Commodification-of-Industrial-Software%3A-A-Case-for-Linden-Lundell/54d6cb77a86e292ff1845eb910c1a1f258e6cee3) for the one used in the sketch, and copied it to `/assets`.

Also added a link to this PR to the pattern, so that anybody working on this in the future can easily find the open conversations in here.

**Pull request merged**
